lib/SILGen/SILGenDestructor.cpp

```diff
@@ -31,15 +31,60 @@ void SILGenFunction::emitDestroyingDestructor(DestructorDecl *dd) {
     Loc.markAutoGenerated();
 
   auto cd = cast<ClassDecl>(dd->getDeclContext());
+  auto &C = cd->getASTContext();
   SILValue selfValue = emitSelfDecl(dd->getImplicitSelfDecl());
 
   // Create a basic block to jump to for the implicit destruction behavior
   // of releasing the elements and calling the superclass destructor.
   // We won't actually emit the block until we finish with the destructor body.
   prepareEpilog(None, false, CleanupLocation(Loc));
 
+  auto cleanupLoc = CleanupLocation(Loc);
+
+  SILBasicBlock *deinitBodyBB = nullptr;
+  SILBasicBlock *finishBB = nullptr;
+  if (cd->isDistributedActor()) {
+    auto remoteBB = createBasicBlock("remoteActorDeinitBB");
+    finishBB = createBasicBlock("finishDeinitBB");
+    deinitBodyBB = createBasicBlock("deinitBodyBB");
+
+    // FIXME: what should the type of management be for this?
+    auto managedSelf = ManagedValue::forBorrowedRValue(selfValue);
+
+    auto selfTy = F.mapTypeIntoContext(cd->getDeclaredInterfaceType());
+    emitDistributedIfRemoteBranch(
+        SILLocation(Loc),
+        managedSelf, selfTy,
+        /*if remote=*/remoteBB,
+        /*if local=*/deinitBodyBB);
+
+    {
+      B.emitBlock(remoteBB);
+
+      // Note that we do NOT execute user-declared the deinit body.
+      // They would be free to access state which does not exist in a remote DA
+
+      // we are a remote instance,
+      // the only properties we can destroy are the id and system properties.
+      for (VarDecl *vd : cd->getStoredProperties()) {
+        if (getActorIsolation(vd) == ActorIsolation::ActorInstance)
+          continue;
+
+        // Just to double-check, we only want to destroy `id` and `actorSystem`
+        if (vd->getBaseIdentifier() == C.Id_id ||
+            vd->getBaseIdentifier() == C.Id_actorSystem) {
+          destroyClassMember(cleanupLoc, managedSelf, vd);
+        }
+      }
+
+      B.createBranch(SILLocation(Loc), finishBB);
+    }
+  }
+
   emitProfilerIncrement(dd->getTypecheckedBody());
```

Comment:
IIUC, this can be emitted into `remoteBB` after branch instruction, creating incorrect SIL. Should it be after `B.emitBlock(deinitBodyBB)`.

Would be nice to have a unit test for deinit and profiling.